docs/data_collection_fundamentals/devtools_continued.md

```diff
@@ -0,0 +1,65 @@
+---
+title: DevTools continued
+description: Continue learning how to collect data from a website using browser DevTools, CSS selectors and JavaScript.
+menuWeight: 20.3
+paths:
+    - data-collection-fundamentals/devtools-continued
+---
+
+# [](#finish-devtools) Finish with DevTools
+
+In the first part of the DevTools tutorial, we were able to collect information about a single website from the <a href="https://www.alexa.com/topsites" target="_blank">Alexa Top Sites index</a>. If you missed the previous part, we suggest going through it to understand the basic concepts, but you don't need any of the code from there, we will start from scratch.
+
+## [](#looping) Looping over elements
+
+> If you're not familiar with the concept of loops in programming, <a href="https://javascript.info/while-for" target="_blank">check out this tutorial</a>. If you're not familiar with functions that iterate over an array, <a href="https://javascript.info/array-methods#iterate-foreach" target="_blank">this tutorial will help</a>.
+
+First, we will use the `querySelectorAll()` function from the previous chapter to get an array of all the elements that contain the sites' data.
+
+```js
+const sites = document.querySelectorAll('div.site-listing');
+console.log(sites);
+```
+
+![Print all websites]({{@asset data_collection_fundamentals/images/print-all-websites.webp}})
+
+Second, we will loop over the sites and print their unstructured data. The `forEach` function accepts one argument, which is a function that will be invoked for each element of the `sites` array.
+
+```js
+sites.forEach((site) => console.log(site.textContent));
+```
+
+> If you're not familiar with arrow functions (`=>`), visit <a href="https://javascript.info/arrow-functions-basics" target="_blank">this tutorial</a> to learn more.
+
+![Print all websites' text]({{@asset data_collection_fundamentals/images/print-all-websites-text.webp}})
+
+## [](#collecting-data) Collecting data in a loop
+
+In the previous chapter we created a `result` object that contained data of a single website. Now we'll create an array of those result objects which will contain all the 50 websites' data. To do that, we'll use a function called `map()` and the collection skills we learned in the previous chapter.
+
+> <a href="https://javascript.info/array-methods#map" target="_blank">Learn more</a> about the `array.map()` function.
+
+```js
+const results = Array.from(sites).map((site) => {
+    const fields = site.querySelectorAll('div.td');
+    return {
+        rank: fields[0].textContent.trim(),
+        site: fields[1].textContent.trim(),
+        dailyTimeOnSite: fields[2].textContent.trim(),
+        dailyPageViews: fields[3].textContent.trim(),
+        percentFromSearch: fields[4].textContent.trim(),
+        totalLinkingSites: fields[5].textContent.trim(),
+    };
+});
+console.log(results);
```

Comment:
Maybe for this and previous example `for (const ... of ...)` loop would be easier. You don't need arrow function (or functions at all), you don't need `Array.from` (elements are iterable), you don't need to understand difference between `.map` and `.forEach`. Also when we would add async later, it naturally works with it.

Reply:
💡